### PR TITLE
Handle API promotion requests and expose them as Tasks

### PR DESCRIPTION
**Pull request by gaetanmaisse**
**Issue**

https://github.com/gravitee-io/issues/issues/5527

**Description**
 
Handle API promotion requests and expose them as Tasks

### CODE REVIEW

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TaskServiceImpl.java [hunk 1]
```diff
@@ -237,6 +251,16 @@ public Metadata getMetadata(List<TaskEntity> tasks) {
...
+                } else if (data instanceof PromotionEntity) {
+                    final PromotionEntity promotion = (PromotionEntity) data;
+                    try {
+                        final Api api = objectMapper.readValue(promotion.getApiDefinition(), Api.class);
+                        final boolean apiAlreadyExist = apiService.exists(api.getId());
+                        metadata.put(promotion.getId(), "apiName", api.getName());
+                        metadata.put(promotion.getId(), "isAnUpdate", apiAlreadyExist);
+                    } catch (JsonProcessingException e) {
+                        LOGGER.error("Unexpected error while getting API definition", e);
+                    }
```
Comment (gaetanmaisse) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TaskServiceImpl.java:263]: To remove as it's unneeded, all the info are in the task directly

File: gravitee-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/promotion/PromotionTasksServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    private PromotionTasksService cut;
+
+    @Before
+    public void setUp() {
+        cut = new PromotionTasksServiceImpl(userService, promotionService, permissionService, environmentService, apiService, objectMapper);
+    }
+
+    @Test
+    public void shouldNotGetPromotionTasksWhenNothingFromDb() {
+        when(userService.findByIdWithRoles(any())).thenReturn(getAUserEntity());
```
Comment (phiz71) [inline gravitee-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/promotion/PromotionTasksServiceImplTest.java:85]: mock `environmentService` instead

File: gravitee-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/promotion/PromotionTasksServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        when(permissionService.hasPermission(eq(RolePermission.ENVIRONMENT_API), any(String.class), eq(CREATE), eq(UPDATE)))
+            .thenReturn(true);
+
+        final List<TaskEntity> result = cut.getPromotionTasks("user");
+        assertThat(result).isEmpty();
+    }
+
+    @Test
+    public void shouldNotGetPromotionTasksIfDoesNotHavePermissions() {
+        when(userService.findByIdWithRoles(any())).thenReturn(getAUserEntity());
```
Comment (phiz71) [inline gravitee-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/promotion/PromotionTasksServiceImplTest.java:97]: mock environmentService instead

File: gravitee-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/promotion/PromotionTasksServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        when(permissionService.hasPermission(eq(RolePermission.ENVIRONMENT_API), any(String.class), eq(CREATE), eq(UPDATE)))
+            .thenReturn(false);
+
+        final List<TaskEntity> result = cut.getPromotionTasks("user");
+        assertThat(result).isEmpty();
+    }
+
+    @Test
+    public void shouldGetPromotionTasks() throws JsonProcessingException {
+        when(userService.findByIdWithRoles(any())).thenReturn(getAUserEntity());
```
Comment (phiz71) [inline gravitee-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/promotion/PromotionTasksServiceImplTest.java:108]: mock environmentService instead

File: gravitee-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/promotion/PromotionTasksServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+    private UserEntity getAUserEntity() {
+        UserEntity user = new UserEntity();
+        final HashMap<String, Set<UserRoleEntity>> envRoles = new HashMap<>();
+        envRoles.put("DEFAULT", Collections.singleton(new UserRoleEntity()));
+        envRoles.put("DEV", Collections.singleton(new UserRoleEntity()));
+        user.setEnvRoles(envRoles);
+
+        return user;
+    }
```
Comment (phiz71) [inline gravitee-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/promotion/PromotionTasksServiceImplTest.java:134]: useless

**Review by phiz71 — CHANGES_REQUESTED** [review]

**Review by phiz71 — APPROVED** [review]

**Pull request merged**